pkg/storage/export/export/export.go

```diff
@@ -1104,3 +1255,129 @@ func (ctrl *VMExportController) pvcConditionFromPVC(pvcs []*corev1.PersistentVol
 	}
 	return cond
 }
+
+func (ctrl *VMExportController) expandVirtualMachine(vm *virtv1.VirtualMachine) (*virtv1.VirtualMachine, error) {
+	instancetypeSpec, err := ctrl.instancetypeMethods.FindInstancetypeSpec(vm)
+	if err != nil {
+		return nil, err
+	}
+	preferenceSpec, err := ctrl.instancetypeMethods.FindPreferenceSpec(vm)
+	if err != nil {
+		return nil, err
+	}
+
+	if instancetypeSpec == nil && preferenceSpec == nil {
+		return vm, nil
+	}
+
+	conflicts := ctrl.instancetypeMethods.ApplyToVmi(field.NewPath("spec", "template", "spec"), instancetypeSpec, preferenceSpec, &vm.Spec.Template.Spec)
+	if len(conflicts) > 0 {
+		return nil, fmt.Errorf("cannot expand instancetype to VM, due to %d conflicts", len(conflicts))
+	}
+
+	// Remove InstancetypeMatcher and PreferenceMatcher, so the returned VM object can be used and not cause a conflict
+	vm.Spec.Instancetype = nil
+	vm.Spec.Preference = nil
+
+	return vm, nil
+}
+
+func (ctrl *VMExportController) updateHttpSourceDataVolumeTemplate(vm *virtv1.VirtualMachine) *virtv1.VirtualMachine {
+	for _, volume := range vm.Spec.Template.Spec.Volumes {
+		volumeName := ""
+		if volume.DataVolume != nil {
+			volumeName = volume.DataVolume.Name
+		}
+		if volume.PersistentVolumeClaim != nil {
+			volumeName = volume.PersistentVolumeClaim.ClaimName
+		}
+		if volumeName != "" {
+			vm.Spec.DataVolumeTemplates = ctrl.replaceUrlDVTemplate(volumeName, vm.Namespace, vm.Spec.DataVolumeTemplates)
+		}
+	}
+	return vm
+}
+
+func (ctrl *VMExportController) replaceUrlDVTemplate(volumeName, namespace string, templates []virtv1.DataVolumeTemplateSpec) []virtv1.DataVolumeTemplateSpec {
+	res := make([]virtv1.DataVolumeTemplateSpec, 0)
+	for _, template := range templates {
+		if template.ObjectMeta.Name == volumeName {
+			//Replace template
+			replacement := template.DeepCopy()
+			replacement.Spec.Source = &cdiv1.DataVolumeSource{
+				HTTP: &cdiv1.DataVolumeSourceHTTP{
+					URL: "",
+				},
+			}
+			res = append(res, *replacement)
+		} else {
+			res = append(res, template)
+		}
+	}
+	return res
+}
+
+func (ctrl *VMExportController) generateVMDefinitionFromVm(vm *virtv1.VirtualMachine) ([]byte, error) {
+	expandedVm, err := ctrl.expandVirtualMachine(vm)
+	if err != nil {
+		return nil, err
+	}
+	// Clear status
+	expandedVm.Status = virtv1.VirtualMachineStatus{}
```

Comment:
Should also clear out metadata fields like uid, resourceversion, creationtimestamp, etc

Reply:
Done